<!DOCTYPE html><head><meta charset="utf-8"/><style> .class-name { color: blue; }  .info-label { font-weight: bold; } .method-section {margin-left: 50px; padding-bottom: 50px;} .method-name { font-weight: bold; color: blue;} .attr-name{color : maroon;} .param-name{color: darkcyan;} .return-name{color: green;} </style></head><body>
<div>
<hr><h1 data-search="enricher" class="class-name">Class Enricher</h1><hr><span class="info-label">Description:</span><br/><br/><span class="class-description">
This class enriches the results of Machaon with third-party information.
</span><br/><br/><h2 class="methods-label">Methods:</h2><br/><br/>
<div data-search="__init__" class="method-section"><span class="method-name" data-search="__init__">__init__()</span><br/><br/><span class="info-label">Description:</span><br/><br/><span class="method-description">
Constructor
</span><br/><br/>
<br/><span class="info-label">Class Attributes:</span><ul>
<li data-search="file_name"><span class="attr-name">file_name [default: '']</span>: <span class="attr-description">
CSV file path for the selected candidate entries. (string)
</span></li>
<li data-search="go_cache"><span class="attr-name">go_cache [default: defaultdict()]</span>: <span class="attr-description">
Gene Ontology cache. Its organization is based on a dictionary with GO term ID as key and a tuple (&lt;GO term&gt;, &lt;GO term type&gt;) as a value
</span></li>
<li data-search="_go_request_url"><span class="attr-name">_go_request_url [default: 'https://www.ebi.ac.uk/QuickGO/services/ontology/go/terms/GO:']</span>: <span class="attr-description">
EBI QuickGO Web Service URL (string)
</span></li>
<li data-search="_enriched_headers"><span class="attr-name">_enriched_headers [default: ['pdbId', 'chainId', 'chainLength', 'resolution', 'geneId', 'taxonomy', 'molecularFunction', 'cellularComponent', 'biologicalProcess']</span>: <span class="attr-description">
Output CSV headers (string)
</span></li>
<li data-search="_request_chunk_size"><span class="attr-name">_request_chunk_size [default: 100]</span>: <span class="attr-description">
Maximum PDB IDs included in an NCBI Entrez Query
</span></li>
<li data-search="root_disk"><span class="attr-name">root_disk [default: '']</span>: <span class="attr-description">
Current setting of root disk where the third-party data and the target dataset reside. (string)
</span></li>
<li data-search="pdb_dataset_path"><span class="attr-name">pdb_dataset_path [default: '']</span>: <span class="attr-description">
PDB directory path (relative) (string)
</span></li>
<li data-search="_uniprot_data_path"><span class="attr-name">_uniprot_data_path [default: '']</span>: <span class="attr-description">
UniProt data cache directory (see set_root_disk()) (string)
</span></li>
<li data-search="_pdb_info_path"><span class="attr-name">_pdb_info_path [default: '']</span>: <span class="attr-description">
PDB meta-data cache directory (see set_root_disk()) (string)
</span></li>
<li data-search="_enrichment_cache"><span class="attr-name">_enrichment_cache [default: '']</span>: <span class="attr-description">
Enrichment data cache directory (see set_root_disk()) (string)
</span></li>
<li data-search="_rcsb_enrichment_cache"><span class="attr-name">_rcsb_enrichment_cache [default: '']</span>: <span class="attr-description">
RCSB enrichment data cache directory (see set_root_disk()) (string)
</span></li>
<li data-search="_entrez_cache"><span class="attr-name">_entrez_cache [default: '']</span>: <span class="attr-description">
NCBI Entez data cache directory (see set_root_disk()) (string)
</span></li>
<li data-search="_uniprot_request_url"><span class="attr-name">_uniprot_request_url [default: 'https://www.uniprot.org/uniprot/ACCESSION.xml']</span>: <span class="attr-description">
UniProt request URL with placeholder ("ACCESSION")
</span></li>
<li data-search="cores"><span class="attr-name">cores [default: 1]</span>: <span class="attr-description">
User-specified maximum processes (int)
</span></li>
<li data-search="verbose"><span class="attr-name">verbose [default: True]</span>: <span class="attr-description">
Toggle verbose output. (boolean)
</span></li>
<li data-search="debugging"><span class="attr-name">debugging [default: False]</span>: <span class="attr-description">
Toggle debugging mode (single core execution). (boolean)
</span></li>
</ul>
</div>
<div data-search="set_root_disk" class="method-section"><span class="method-name" data-search="set_root_disk">set_root_disk()</span><br/><br/><span class="info-label">Description:</span><br/><br/><span class="method-description">
Sets all the required directories according to the provided root directory path.
</span><br/><br/>
<br/><span class="info-label">Arguments:</span><ul>
<li data-search="root_disk"><span class="param-name">root_disk</span>: <span class="arg-description">
Directory path where the third-party data and the target dataset reside. (string)
</span></li>
</ul>
</div>
<div data-search="load_go_cache" class="method-section"><span class="method-name" data-search="load_go_cache">load_go_cache()</span><br/><br/><span class="info-label">Description:</span><br/><br/><span class="method-description">
Loads the Gene Ontology cache CSV file into a dictionary (class attribute self.go_cache)
</span><br/><br/>
</div>
<div data-search="update_go_cache" class="method-section"><span class="method-name" data-search="update_go_cache">update_go_cache()</span><br/><br/><span class="info-label">Description:</span><br/><br/><span class="method-description">
Updates the Gene Ontology cache with new data.
</span><br/><br/>
<br/><span class="info-label">Arguments:</span><ul>
<li data-search="term_id"><span class="param-name">term_id</span>: <span class="arg-description">
GO Term ID (string)
</span></li>
<li data-search="term_info"><span class="param-name">term_info</span>: <span class="arg-description">
Tupple with (GO term, Go term type) value, e.g. ('angiogenesis', 'biologicalProcess')
</span></li>
</ul>
</div>
<div data-search="get_uniprot_data" class="method-section"><span class="method-name" data-search="get_uniprot_data">get_uniprot_data()</span><br/><br/><span class="info-label">Description:</span><br/><br/><span class="method-description">
Retrieves XML formatted information for a UniProt Accession identifier.
</span><br/><br/>
<br/><span class="info-label">Arguments:</span><ul>
<li data-search="uniprot_accession"><span class="param-name">uniprot_accession</span>: <span class="arg-description">
UniProt Accession identifier (string)
</span></li>
</ul>
<span class="info-label">Returns:</span><ul>
<li data-search="web_data"><span class="return-name">web_data</span>: <span class="return-description">
XML formatted request result (string)
</span></li>
</ul>
</div>
<div data-search="access_pdb_info" class="method-section"><span class="method-name" data-search="access_pdb_info">access_pdb_info()</span><br/><br/><span class="info-label">Description:</span><br/><br/><span class="method-description">
Retrieves meta-data for a PDB ID.
</span><br/><br/>
<br/><span class="info-label">Arguments:</span><ul>
<li data-search="entry"><span class="param-name">entry</span>: <span class="arg-description">
Composite input: PDB ID (string), PDB chain ID (string)
</span></li>
</ul>
<span class="info-label">Returns:</span><ul>
<li data-search="pdb_id"><span class="return-name">pdb_id</span>: <span class="return-description">
PDB ID (string)
</span></li>
<li data-search="residues"><span class="return-name">residues</span>: <span class="return-description">
PDB chain length (int)
</span></li>
<li data-search="resolution"><span class="return-name">resolution</span>: <span class="return-description">
resolution in angstrom (Ã…) (float)
</span></li>
<li data-search="organism"><span class="return-name">organism</span>: <span class="return-description">
origin of protein (string)
</span></li>
<li data-search="gene_names"><span class="return-name">gene_names</span>: <span class="return-description">
gene name (string)
</span></li>
</ul>
</div>
<div data-search="retrieve_uniprot_mapping" class="method-section"><span class="method-name" data-search="retrieve_uniprot_mapping">retrieve_uniprot_mapping()</span><br/><br/><span class="info-label">Description:</span><br/><br/><span class="method-description">
Retrieves UniProt Accession identifier.
</span><br/><br/>
<br/><span class="info-label">Arguments:</span><ul>
<li data-search="entry"><span class="param-name">entry</span>: <span class="arg-description">
Composite input: PDB ID, PDB chain ID
</span></li>
</ul>
<span class="info-label">Returns:</span><ul>
<li data-search="uniprot_accession"><span class="return-name">Structure ID</span>: <span class="return-description">
Concatenated and underscore separated PDB ID and PDB chain ID. (string)
</span></li>
<li data-search="uniprot_accession"><span class="return-name">UniProt identifier</span>: <span class="return-description">
UniProt Accession identifier (string)
</span></li>
</ul>
</div>
<div data-search="fetch_enrichment" class="method-section"><span class="method-name" data-search="fetch_enrichment">fetch_enrichment()</span><br/><br/><span class="info-label">Description:</span><br/><br/><span class="method-description">
Enriches the results with Gene Ontology, UniProt, RCSB and Entrez information. All the data are cached.
</span><br/><br/>
<br/><span class="info-label">Arguments:</span><ul>
<li data-search="data"><span class="param-name">data</span>: <span class="arg-description">
Pandas dataframe
</span></li>
<li data-search="column_headers"><span class="param-name">column_headers [default: ['b-phipsi', 'w-rdist', 't-alpha']</span>: <span class="arg-description">
Headers of the columns that will be included in the output result except for structure ID. (list of strings)
</span></li> 
</ul>
</div>
<div data-search="read_enrichment" class="method-section"><span class="method-name" data-search="read_enrichment">read_enrichment()</span><br/><br/><span class="info-label">Description:</span><br/><br/><span class="method-description">
Reads information from local static UniProt resources.
</span><br/><br/>
<br/><span class="info-label">Arguments:</span><ul>
<li data-search="pdb_id"><span class="param-name">pdb_id</span>: <span class="arg-description">
PDB ID (string)
</span></li>
</ul>
<span class="info-label">Returns:</span><ul>
<li data-search="enrichment"><span class="return-name">enrichment</span>: <span class="return-description">
Comma-separated enrichment information : &lt;Entrez Gene ID&gt;, &lt;NCBI Taxon ID&gt;, &lt;Gene Ontology molecular function terms&gt; <br/>
&lt;Gene Ontology cellular component terms&gt;, &lt;Gene Ontology biological process terms&gt; (string) ("#" for each missing information)
</span></li>
<li data-search="pdb_info"><span class="return-name">pdb_info</span>: <span class="return-description">
Dictionary with PDB meta-data (see get_pdb_info())
</span></li>
<li data-search="pdb_id"><span class="return-name">pdb_id</span>: <span class="return-description">
The PDB ID that the requested information belongs to (for multi-processing purposes). (string)
</span></li>
</ul>
</div>
<div data-search="map_uniprot_to_entrez" class="method-section"><span class="method-name" data-search="map_uniprot_to_entrez">map_uniprot_to_entrez()</span><br/><br/><span class="info-label">Description:</span><br/><br/><span class="method-description">
Retrieves Entrez Gene IDs for UniProt Accession identifiers
</span><br/><br/>
<br/><span class="info-label">Arguments:</span><ul>
<li data-search="accession_ids"><span class="param-name">accession_ids</span>: <span class="arg-description">
Accession IDs (list of strings)
</span></li>
</ul>
<span class="info-label">Returns:</span><ul>
<li data-search="results"><span class="return-name">results</span>: <span class="return-description">
Entrez Gene IDs ("#" for each missing ID) (list of strings)
</span></li>
</ul>
</div>
<div data-search="split_go_annotations" class="method-section"><span class="method-name" data-search="split_go_annotations">split_go_annotations()</span><br/><br/><span class="info-label">Description:</span><br/><br/><span class="method-description">
Groups Gene Ontology annotations by term type
</span><br/><br/>
<br/><span class="info-label">Arguments:</span><ul>
<li data-search="terms"><span class="param-name">terms</span>: <span class="arg-description">
GO term IDs
</span></li>
</ul>
<span class="info-label">Returns:</span><ul>
<li data-search="data"><span class="return-name">data</span>: <span class="return-description">
Terms of the same type are concatenated with "#" separator. One string per type (list of strings)
</span></li>
</ul>
</div>
<div data-search="get_enrichment_info" class="method-section"><span class="method-name" data-search="get_enrichment_info">get_enrichment_info()</span><br/><br/><span class="info-label">Description:</span><br/><br/><span class="method-description">
Retrieve enrichment information from RCSB. All the data are cached. (also see read_enrichment(), this method is a fallback.)
</span><br/><br/>
<br/><span class="info-label">Arguments:</span><ul>
<li data-search="structure_ids"><span class="param-name">structure_ids</span>: <span class="arg-description">
Composite identifier "&lt;PDB ID&gt;_&lt;PDB chain ID&gt;"
</span></li>
</ul>
<span class="info-label">Returns:</span><ul>
<li data-search="results"><span class="return-name">results</span>: <span class="return-description">
Enrichment information (same as in read_enrichment()) (list of strings)
</span></li>
</ul>
</div>
<div data-search="get_go_term" class="method-section"><span class="method-name" data-search="get_go_term">get_go_term()</span><br/><br/><span class="info-label">Description:</span><br/><br/><span class="method-description">
Retrieves information about a Gene Ontology term ID from EBI QuickGO.
</span><br/><br/>
<br/><span class="info-label">Arguments:</span><ul>
<li data-search="go_id"><span class="param-name">go_id</span>: <span class="arg-description">
Term ID (string)
</span></li>
</ul>
<span class="info-label">Returns:</span><ul>
<li data-search="go_term"><span class="return-name">go_term</span>: <span class="return-description">
Term name (string)
</span></li>
<li data-search="term_type"><span class="return-name">term_type</span>: <span class="return-description">
Term type (string)
</span></li>
</ul>
</div>
</div></body>
